ddtrace/profiling/collector/memalloc.py

```diff
@@ -100,48 +103,87 @@ def _get_thread_id_ignore_set(self):
 
     def snapshot(self):
         thread_id_ignore_set = self._get_thread_id_ignore_set()
-        return (
-            tuple(
-                MemoryHeapSampleEvent(
-                    thread_id=thread_id,
-                    thread_name=_threading.get_thread_name(thread_id),
-                    thread_native_id=_threading.get_thread_native_id(thread_id),
-                    frames=stack,
-                    nframes=nframes,
-                    size=size,
-                    sample_size=self.heap_sample_size,
+
+        stacks = [
+            ((stack, nframes, thread_id), size)
+            for (stack, nframes, thread_id), size in _memalloc.heap()
+            if not self.ignore_profiler or thread_id not in thread_id_ignore_set
+        ]
+        if self.export_libdd_enabled:
+            for (frames, nframes, thread_id), size in stacks:
+                ddup.start_sample(nframes)
```

Comment:
do you reckon that we could make this look like a context manager from the Python side, so that we could write something like
~~~
with sample as ddup.sample():
    sample.push_heap(size)
    ...
~~~

Reply:
Ah, yes--that's an interesting idea.

There are several ways of doing this and I think we can make it quite ergonomic, but

1. The native code actually hardcodes precisely two such contexts and switch between them.  This is a poor man's way of implementing async uploads.  So the context we send would be more like a view into a context, rather than the creation of a new context.

2. I'm thinking more and more of moving as much of the profiler into native code.  If that's an agreeable goal, then I would have a slight preference not to invest too much in ergonomics here at this point

With that in mind, could I offer you some à la carte options, sir?
1.  Ignore for now.  Either we fix in a different PR or we never fix it, TBD
2.  Fix for this PR (since when else are we gonna do it?)

Reply:
I'm fine with making this potential follow-up work and keep the current approach for now.